MM-40581: Sanitized user and added all information of a user to publish

nongpond03 opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
Sanitized user and added all information of a user to publish.

If a tester wants to know about impacted APIs I can send information I already checked it. 
-->

#### Ticket Link
<!--
https://github.com/mattermost/mattermost-server/issues/19362
-->

#### Release Note
<!--
Added all information of a user to publish when user_added event is triggered.
-->

mattermost-build commented:
Hello @nongpond03,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

Per the [Mattermost Contribution Guide](https://developers.mattermost.com/contribute/getting-started/), we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**
Once you have signed the CLA, please comment with `/check-cla` and confirm that the CLA check is green.

This is a standard procedure for many open source projects.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

nongpond03 commented:
/check-cla

agarciamontoro reviewed (commented):
Looking at the `Sanitize` code, I wonder whether the email should be also removed (for which we should pass a map containing an `"email": false` entry, if I'm not mistaken), and I don't know whether removing it for everyone, including the user that triggered the event, makes sense. I'll defer to @larkox\s review, who better knows what the client will need at this point.

nongpond03 commented:
Thank your sir. Moreover, Ii seems like the code can support removing "email", "fullname", "passwordupdate" and "authservice". If you want to remove any field just let me know.

thank you sir

larkox commented on app/channel.go:
```diff
@@ -186,7 +186,40 @@ func (a *App) JoinDefaultChannels(c request.CTX, teamID string, user *model.User
 		message := model.NewWebSocketEvent(model.WebsocketEventUserAdded, "", channel.Id, "", nil, "")
+		user.Sanitize(map[string]bool{})
 		message.Add("user_id", user.Id)
+		message.Add("user_create_at", user.CreateAt)
```
Instead of passing every single property, I would just pass a "user" property with the whole user.

larkox reviewed (changes requested):
I would approach this a bit differently (and simpler).

agarciamontoro commented:
> I think that is something regarding the server configuration, right? So it should handle by the sanitize function, right?

The argument passed to the `user.Sanitize` function lets you specify the specific behaviour: https://github.com/mattermost/mattermost-server/blob/master/model/user.go#L644-L663

larkox commented:
Maybe in this case we should use `SanitizeProfile`, along with the `a.Config().GetSanitizeOptions()` as it is used in other places of the app. But looking at that, I am worried the way it is done is prone to errors like this one.

agarciamontoro commented:
That makes more sense, indeed, I was not aware of that function.

nongpond03 commented:
I will change it and make a pull request again.

thank you for your support.

nongpond03 commented:
I already modified it.

First: I changed from passing every single property of user model to be passing in a 'user' property only.
Second: I changed the way sanitization from using Sanitize function to be SanitizeProfile and get config from a.Config().GetSanitizeOptions() to use with it.

If there is anything wrong with the code you can tag me I will come back and fix it as fast as possible.

thank you for your support.

larkox commented on app/channel.go:
```diff
@@ -185,8 +185,11 @@ func (a *App) JoinDefaultChannels(c request.CTX, teamID string, user *model.User
+		user.SanitizeProfile(options)
+
 		message := model.NewWebSocketEvent(model.WebsocketEventUserAdded, "", channel.Id, "", nil, "")
-		message.Add("user_id", user.Id)
```
In order to be backwards compatible, keep the `user_id` also in the message. So the message will have:
`user_id`, `user` and `team_id`.

nongpond03 replied:
Done

Thank you for your support @larkox

larkox reviewed (changes requested):
One minor change and should be good to go from my side :D

larkox commented:
/check-cla

larkox commented:
@nongpond03 Would you be able to sign the CLA? The information is in this comment: https://github.com/mattermost/mattermost-server/pull/22209#issuecomment-1411234091

nongpond03 commented:
/check-cla

nongpond03 commented:
Let me check.

Thank you for helping me @larkox

nongpond03 commented:
/check-cla

nongpond03 commented:
"api/v4/teams/{team_id:[A-Za-z0-9]+}/import" POST (team.go) api4
"api/v4/teams" POST (team.go) api4
"api/v4/users" POST (user.go) api4
"api/v4/teams/{team_id:[A-Za-z0-9]+}/members/batch" POST api4
"api/v4/teams/{team_id:[A-Za-z0-9]+}/members" POST api4
"api/v4/teams/members/invite" POST api4
"/login/{service:[A-Za-z0-9]+}/complete"
"api/v4/teams/members/invite" POST api4
"api/v4/users" POST (user.go) api4
"/promote" (user_local.go)
"api/v4/teams/members/invite" POST api4
"api/v4/teams/{team_id:[A-Za-z0-9]+}/import" (team.go) api4
"api/v4/channels/{channel_id:[A-Za-z0-9]+}/members/" api4

Above apis are impacted. Just in case if you want to do load test.

Thank you

nongpond03 commented:
@larkox Hi, I am not sure that this task is done, isn't it? I want to do another task Can I ?

agarciamontoro commented:
Hey, @nongpond03, sorry for the delay! This is in my hands now, since I want to run a load test and make sure this does not introduce any performance regression. Please go ahead and take any other ticket :)

nongpond03 commented:
> @nongpond03 While we finish reviewing this, feel free to take any other ticket. As long as you feel confident that you will have enough bandwidth to address potential reviews and request for changes, feel free to take as many tickets as you feel capable. 😄

okay it is time to look for another ticket.

thank you

nongpond03 commented:
> Hey, @nongpond03, sorry for the delay! This is in my hands now, since I want to run a load test and make sure this does not introduce any performance regression. This may take a bit of time, so please go ahead and take any other ticket :)

okay

thank you

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

larkox commented:
@agarciamontoro friendly reminder of this PR

agarciamontoro commented:
Thanks for the reminder! I'll try to prioritize it this week :)

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

larkox commented:
@agarciamontoro Friendly reminder of this

@nongpond03 Sorry for the delay, but could you address the merge conflicts?

nongpond03 commented:
> @agarciamontoro Friendly reminder of this
> 
> @nongpond03 Sorry for the delay, but could you address the merge conflicts?



> > @agarciamontoro Friendly reminder of this
> > @nongpond03 Sorry for the delay, but could you address the merge conflicts?
> 
> I will come back to check it today sir.

I think I already merged conflicts.

if I miss something I will come back to fix it as fast as I can.

Thank you for helping me

larkox commented on server/channels/app/channel.go:
```diff
@@ -187,6 +187,7 @@ func (a *App) JoinDefaultChannels(c request.CTX, teamID string, user *model.User
 
 		message := model.NewWebSocketEvent(model.WebsocketEventUserAdded, "", channel.Id, "", nil, "")
 		message.Add("user_id", user.Id)
+		message.Add("user", user)
```
Seems like the user sanitization has been removed. Why?

nongpond03 replied:
sorry my bad let me check again.

nongpond03 replied:
I have fixed it. Sorry for my carelessness. 

Because I found out that the "channel.go" file was deleted from redesigning structure.  I was afraid of copying the whole old file to the new "channel.go" file.

by the way, I am so sorry again.

larkox replied:
No problem. To err is human :smile: That is what the code reviews are for.

nongpond03 replied:
thank you, guys. I appreciate it.

One more question.
Now, I am coding this task "https://github.com/mattermost/mattermost-server/issues/21193" "React"+"Golang". So, I have to move to code in "mattermost-server" repo, don't I ?

Thank you again.

larkox replied:
@nongpond03 Yes. Webapp and Server code now live in the same repo. So you can code the golang part and the react part in one single repo (`mattermost-server`).

agarciamontoro commented:
The initial tests show a generalized increase in both API and store times, probably caused by the increase in the websocket event size. I'm running an additional test to make sure this is not some oddity introduced by the test and will investigate further, but I wanted to ask something in the meantime:
I understand that the goal of adding the user in this event is to avoid making additional requests in the client. I assume that after we modify the client to eliminate those calls, this modification will be a net positive, but only adding more data to the websocket will definitely make things worse in the interim. Have we considered a holistic approach for this in which we not only add the data to the event, but we modifiy the webapp to reduce the number of requests?

agarciamontoro commented:
I ran another pair of tests with and without the modification in this PR, and I found no significant changes this time. In total, I did four bounded tests: two with the changes, two without the changes. Of those, the one that showed the worst peak of times was actually one without the changes. Sadly, the conclusion here is that we need some more investigation.
The concern I have is that testing this is slightly more complex, since what we are interested in is the period of time when users are logging in, since that's where the changes in this PR may make things worse. However, that period of time is always more noisy, which makes it harder to tell whether we are seeing is something relevant or not.
Before merging, I'd like us to

agarciamontoro commented:
On another note: I'll be out on vacation starting tomorrow. I will be back on April 10. So let me ping @agnivade here in case you need something from our team during this week. If not, we can continue the conversation when I'm back. Sorry for the push-back on this, and for the delay!

agarciamontoro reviewed (changes requested):
I ran another pair of tests with and without the modification in this PR, and I found no significant changes this time. In total, I did four bounded tests: two with the changes, two without the changes. Of those, the one that showed the worst peak of times was actually one without the changes. Sadly, the conclusion here is that we need some more investigation.
The concern I have is that interpreting the results of those tests is slightly more complex, since what we are interested in is the period of time when users are logging in, as that's where the changes in this PR may make things worse. However, that period of time is always more noisy, which makes it harder to tell whether what we are seeing is something relevant or not. See how different things seem in the four permutations (`after-1` and `after-2` are the tests with the modifications in this PR, while `before-1` and `before-2` are the tests without the modifications) -- It seems that both `after-1` and `before-2` have some noise there:
![avg-api-times](https://user-images.githubusercontent.com/3924815/229231957-37d0dd58-a2f1-4968-8278-9d817651741f.png)

![avg-api-times](https://user-images.githubusercontent.com/3924815/229232010-4b5e11ae-bec2-4c5e-9084-31d37903478f.png)

![avg-api-times](https://user-images.githubusercontent.com/3924815/229232080-e1ccd7a2-cc2e-4b54-9e07-305d16703317.png)


![avg-api-times](https://user-images.githubusercontent.com/3924815/229232052-b6d4a864-26dc-4867-9a9b-ef53157b4a80.png)



Before merging, I'd like us to consider what I said in my previous comment: if we can combine these changes with the corresponding changes in the clients at the same time, I'd be less concerned about the changes. I haven't looked into it in depth, so I need you guys to tell me: how feasible is that?

For the sake of completeness, here's [the Grafana URL for the deployment](http://44.204.240.213:3000/?orgId=1&var-server=agent-0:4000&var-server=agent-0:9100&var-server=agent-1:4000&var-server=agent-1:9100&var-server=agent-2:4000&var-server=agent-2:9100&var-server=agent-3:4000&var-server=agent-3:9100&var-server=agent-4:4000&var-server=agent-4:9100&var-server=app-0:8067&var-server=app-0:9100&var-server=app-1:8067&var-server=app-1:9100&var-server=localhost:9090&var-server=localhost:9100&var-server=metrics:9090&var-server=metrics:9100&var-server=proxy:9100&from=1680208076031&to=1680285218454): the first test is actually `after-1`, the second  is `before-1`, the third is `before-2` and the fourth is `after-2`.

agnivade commented:
> Before merging, I'd like us to consider what I said in my previous comment: if we can combine these changes with the corresponding changes in the clients at the same time, I'd be less concerned about merging it. I haven't looked into it in depth, so I need you guys to tell me: how feasible is that?

I agree, we should do that. @nongpond03 - wondering if you are interested to submit a PR for the webapp for this?

nongpond03 commented:
> > Before merging, I'd like us to consider what I said in my previous comment: if we can combine these changes with the corresponding changes in the clients at the same time, I'd be less concerned about merging it. I haven't looked into it in depth, so I need you guys to tell me: how feasible is that?
> 
> I agree, we should do that. @nongpond03 - wondering if you are interested to submit a PR for the webapp for this?

Okay, I will do it I think I should ask @nevyangelova right ?

nongpond03 commented:
> > Before merging, I'd like us to consider what I said in my previous comment: if we can combine these changes with the corresponding changes in the clients at the same time, I'd be less concerned about merging it. I haven't looked into it in depth, so I need you guys to tell me: how feasible is that?
> 
> I agree, we should do that. @nongpond03 - wondering if you are interested to submit a PR for the webapp for this?

@agnivade @larkox Hi guys, I hope you are well. I don't want to bother you guys. But, can you tell me what exactly you want me to modify it? I know you want to reduce calling api something. That is all.

Thank you for helping me.
Bond

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

agnivade commented:
Hi @nongpond03 - We need to understand what this PR is for. We are adding the user object to the `user_added` websocket event. The reason to do this is so that clients getting the websocket event don't need to make another request to get the user again, and can get it right from the event itself.

So you'd need to look into the webapp code, see how we are handling the user_added event, and instead of making the request, use the user object passed along with the event.
